File: pkg/waffle/cache_test.go
package main

import (
	"testing"
)

func TestCacheOperations(t *testing.T) {
	cache := NewCache([]string{"key1", "key2"}, []string{"value1", "value2"}, 3)

	if cache.Size() != 2 {
		t.Errorf("Expected size 2, got %d", cache.Size())
	}

	if !cache.CheckIfKeyExists("key1") {
		t.Error("key1 should exist")
	}

	if cache.CheckIfKeyExists("key3") {
		t.Error("key3 should not exist")
	}

	if v := cache.GetValue("key1"); v != "value1" {
		t.Errorf("Expected value1, got %s", v)
	}

	if v := cache.GetValueWithoutPositionChange("key2"); v != "value2" {
		t.Errorf("Expected value2, got %s", v)
	}

	cache.InsertIntoCache("key3", "value3")
	if cache.Size() != 3 {
		t.Errorf("Expected size 3, got %d", cache.Size())
	}

	if !cache.CheckIfKeyExists("key3") {
		t.Error("key3 should exist")
	}

	cache.InsertIntoCache("key4", "value4")
	if cache.Size() != 3 {
		t.Errorf("Expected size 3, got %d", cache.Size())
	}

	if cache.CheckIfKeyExists("key1") {
		t.Error("key1 should not exist")
	}

	evictedKey, evictedValue := cache.EvictLRElementFromCache()
	if evictedKey != "key2" || evictedValue != "value2" {
		t.Errorf("Expected evicted (key2, value2), got (%s, %s)", evictedKey, evictedValue)
	}

	if cache.Size() != 2 {
		t.Errorf("Expected size 2, got %d", cache.Size())
	}

	value, isPresent := cache.GetValueWithoutPositionChangeNew("key3")
	if !isPresent {
		t.Error("key3 should be present")
	}
	if value != "value3" {
		t.Errorf("Expected value3, got %s", value)
	}

	value, isPresent = cache.GetValueWithoutPositionChangeNew("key1")
	if isPresent {
		t.Error("key1 should not be present")
	}
	if value != "" {
		t.Errorf("Expected empty string, got %s", value)
	}
}

File: waffle/benchmark/src/cache_test.cpp
#include "Cache.hpp"
#include <string>
#include <vector>
#include <iostream>

int main(){ 
    std::vector<std::string> keys;
    std::vector<std::string> values;

    keys.push_back("K1");
    keys.push_back("K2");
    values.push_back("V1");
    values.push_back("V2");

    Cache cache = Cache(keys, values, 3);

    if (cache.size()!=2){
        std::cout<<"Expected size 2\n";
    }

    if(!cache.checkIfKeyExists("K1")){
        std::cout<<"K1 Should Exist!\n";
    }

    if (cache.checkIfKeyExists("K3")) {
		std::cout<<"K3 Should Not Exist!\n";
	}
    
    if(cache.getValueWithoutPositionChange("K1")!="V1"){
        std::cout<<"Expected V1 !\n";
    }

    cache.insertIntoCache("K3", "V3");
    if (cache.size() != 3) {
		std::cout<<"Expected Size 3\n";
	}

    if (!cache.checkIfKeyExists("K3")) {
		std::cout<<("key3 should exist");
	}

    cache.insertIntoCache("K4", "V4");

	if (cache.size() != 3) {
		std::cout<<"Expected Size 3\n";
	}

    if (cache.checkIfKeyExists("K1")) {
		std::cout<<("key1 should not exist");
	}


    std::vector<std::string> expectd = cache.evictLRElementFromCache();
    if (expectd[0] != "K2" || expectd[1] != "V2") {
		std::cout<<"Expected evicted (K2, V2)\n";
	}

    if (cache.size() != 2) {
		std::cout<<"Expected Size 2\n";
	}

    bool isPresent=false;
    std::string val = cache.getValueWithoutPositionChangeNew("K3",isPresent);
	if (!isPresent) {
		std::cout<<("key3 should be present");
	}
	if (val != "V3") {
		std::cout<<("Expected V3");
	}


    isPresent=false;
    std::string val2 = cache.getValueWithoutPositionChangeNew("K1",isPresent);
	if (isPresent) {
		std::cout<<("key1 should not be present");
	}
	if (val2 != "") {
		std::cout<<("Expected empty string");
	}
    
}